File: src/backend/routers/skills.py
"""
Skills Router - API endpoints for skills management.

Endpoints:
- GET /api/skills/library - List available skills
- GET /api/skills/library/{name} - Get skill content
- POST /api/skills/library/sync - Sync library from GitHub
- GET /api/skills/library/status - Get library status
- GET /api/agents/{name}/skills - List assigned skills
- PUT /api/agents/{name}/skills - Bulk update assignments
- POST /api/agents/{name}/skills/inject - Push skills to running agent
"""

from typing import List
from fastapi import APIRouter, Depends, HTTPException

from models import User
from dependencies import get_current_user, require_admin
from database import db
from db_models import AgentSkill, SkillInfo, AgentSkillsUpdate
from services.skill_service import skill_service

router = APIRouter(prefix="/api", tags=["skills"])


# ============================================================================
# Skills Library Endpoints
# ============================================================================

@router.get("/skills/library", response_model=List[SkillInfo])
async def list_skills(current_user: User = Depends(get_current_user)):
    """
    List all available skills from the skills library.

    Returns skills with name, description, and path.
    Content is not included for performance.
    """
    skills = skill_service.list_skills()
    return [
        SkillInfo(
            name=s["name"],
            description=s.get("description"),
            path=s["path"]
        )
        for s in skills
    ]


@router.get("/skills/library/status")
async def get_library_status(current_user: User = Depends(get_current_user)):
    """
    Get the current status of the skills library.

    Returns configuration status, sync info, and skill count.
    """
    return skill_service.get_library_status()


@router.get("/skills/library/{skill_name}")
async def get_skill(
    skill_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get details for a specific skill including full content.
    """
    skill = skill_service.get_skill(skill_name)
    if not skill:
        raise HTTPException(status_code=404, detail=f"Skill '{skill_name}' not found")
    return skill


@router.post("/skills/library/sync")
async def sync_library(admin_user: User = Depends(require_admin)):
    """
    Sync the skills library from GitHub.

    Admin-only. Clones or pulls the configured repository.
    """
    result = skill_service.sync_library()
    if not result.get("success"):
        raise HTTPException(
            status_code=400,
            detail=result.get("error", "Sync failed")
        )
    return result


# ============================================================================
# Agent Skills Assignment Endpoints
# ============================================================================

@router.get("/agents/{agent_name}/skills", response_model=List[AgentSkill])
async def get_agent_skills(
    agent_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Get skills assigned to an agent.

    Returns list of AgentSkill objects with assignment metadata.
    """
    # TODO: Add access control for shared agents
    return db.get_agent_skills(agent_name)


@router.put("/agents/{agent_name}/skills")
async def update_agent_skills(
    agent_name: str,
    update: AgentSkillsUpdate,
    current_user: User = Depends(get_current_user)
):
    """
    Bulk update skills assigned to an agent.

    Replaces all existing skill assignments with the provided list.
    """
    # TODO: Add access control (owner only)
    count = db.set_agent_skills(
        agent_name=agent_name,
        skill_names=update.skills,
        assigned_by=current_user.username
    )
    return {
        "success": True,
        "agent_name": agent_name,
        "skills_assigned": count,
        "skills": update.skills
    }


# NOTE: inject endpoint MUST be defined BEFORE {skill_name} routes
# to prevent FastAPI from matching "inject" as a skill_name parameter
@router.post("/agents/{agent_name}/skills/inject")
async def inject_skills(
    agent_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Inject assigned skills into a running agent.

    Copies all assigned skills to the agent's .claude/skills/ directory.
    Agent must be running.
    """
    result = await skill_service.inject_skills(agent_name)
    if not result.get("success") and result.get("skills_failed", 0) > 0:
        # Partial success or full failure
        return result

    return result


@router.post("/agents/{agent_name}/skills/{skill_name}")
async def assign_skill(
    agent_name: str,
    skill_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Assign a single skill to an agent.
    """
    # Verify skill exists in library
    skill = skill_service.get_skill(skill_name)
    if not skill:
        raise HTTPException(
            status_code=404,
            detail=f"Skill '{skill_name}' not found in library"
        )

    result = db.assign_skill(agent_name, skill_name, current_user.username)
    if result is None:
        return {
            "success": True,
            "message": "Skill already assigned",
            "skill_name": skill_name
        }

    return {
        "success": True,
        "message": "Skill assigned",
        "skill": result
    }


@router.delete("/agents/{agent_name}/skills/{skill_name}")
async def unassign_skill(
    agent_name: str,
    skill_name: str,
    current_user: User = Depends(get_current_user)
):
    """
    Remove a skill assignment from an agent.
    """
    removed = db.unassign_skill(agent_name, skill_name)
    return {
        "success": True,
        "removed": removed,
        "skill_name": skill_name
    }

File: src/mcp-server/src/tools/skills.ts
/**
 * Skills Management Tools
 *
 * MCP tools for managing Trinity agent skills:
 * - list_skills: List available skills from the library
 * - get_skill: Get skill details and content
 * - assign_skill_to_agent: Assign a skill to an agent
 * - sync_agent_skills: Inject assigned skills to a running agent
 */

import { z } from "zod";
import { TrinityClient } from "../client.js";
import type { McpAuthContext } from "../types.js";

/**
 * Skill information from the library
 */
interface SkillInfo {
  name: string;
  description: string | null;
  path: string;
  content?: string;
}

/**
 * Skills library status
 */
interface SkillsLibraryStatus {
  configured: boolean;
  url: string | null;
  branch: string;
  cloned: boolean;
  last_sync: string | null;
  commit_sha: string | null;
  skill_count: number;
}

/**
 * Skill injection result
 */
interface SkillInjectionResult {
  success: boolean;
  skills_injected: number;
  skills_failed: number;
  results: Record<string, { success: boolean; error?: string }>;
}

/**
 * Create skills management tools with the given client
 */
export function createSkillsTools(
  client: TrinityClient,
  requireApiKey: boolean
) {
  /**
   * Get Trinity client with appropriate authentication
   */
  const getClient = (authContext?: McpAuthContext): TrinityClient => {
    if (requireApiKey) {
      if (!authContext?.mcpApiKey) {
        throw new Error("MCP API key authentication required but no API key found in request context");
      }
      const userClient = new TrinityClient(client.getBaseUrl());
      userClient.setToken(authContext.mcpApiKey);
      return userClient;
    }
    return client;
  };

  return {
    // ========================================================================
    // list_skills - List available skills from library
    // ========================================================================
    listSkills: {
      name: "list_skills",
      description:
        "List all available skills from the skills library. " +
        "Returns skill names, descriptions, and paths. " +
        "Skills are loaded from the configured GitHub repository. " +
        "Use get_skill to get the full content of a specific skill.",
      parameters: z.object({}),
      execute: async (_params: unknown, context?: { session?: McpAuthContext }) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const skills = await apiClient.request<SkillInfo[]>(
          "GET",
          "/api/skills/library"
        );

        if (skills.length === 0) {
          return JSON.stringify({
            message: "No skills available. The skills library may not be configured or synced.",
            hint: "Configure skills_library_url in Settings and sync the library.",
            skills: []
          }, null, 2);
        }

        return JSON.stringify({
          count: skills.length,
          skills: skills.map(s => ({
            name: s.name,
            description: s.description || "No description",
            path: s.path
          }))
        }, null, 2);
      },
    },

    // ========================================================================
    // get_skill - Get skill details and content
    // ========================================================================
    getSkill: {
      name: "get_skill",
      description:
        "Get full details for a specific skill from the library, including its content. " +
        "Use this to see what a skill teaches an agent to do before assigning it.",
      parameters: z.object({
        skill_name: z.string().describe("Name of the skill to retrieve"),
      }),
      execute: async (
        { skill_name }: { skill_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const skill = await apiClient.request<SkillInfo>(
          "GET",
          `/api/skills/library/${encodeURIComponent(skill_name)}`
        );

        return JSON.stringify(skill, null, 2);
      },
    },

    // ========================================================================
    // get_skills_library_status - Get library sync status
    // ========================================================================
    getSkillsLibraryStatus: {
      name: "get_skills_library_status",
      description:
        "Get the current status of the skills library. " +
        "Shows whether it's configured, synced, and how many skills are available.",
      parameters: z.object({}),
      execute: async (_params: unknown, context?: { session?: McpAuthContext }) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const status = await apiClient.request<SkillsLibraryStatus>(
          "GET",
          "/api/skills/library/status"
        );

        return JSON.stringify(status, null, 2);
      },
    },

    // ========================================================================
    // assign_skill_to_agent - Assign a skill to an agent
    // ========================================================================
    assignSkillToAgent: {
      name: "assign_skill_to_agent",
      description:
        "Assign a skill to an agent. " +
        "The skill will be injected when the agent starts, or you can use sync_agent_skills to inject immediately. " +
        "Skills teach agents specific behaviors defined in SKILL.md files.",
      parameters: z.object({
        agent_name: z.string().describe("Name of the agent to assign the skill to"),
        skill_name: z.string().describe("Name of the skill to assign"),
      }),
      execute: async (
        { agent_name, skill_name }: { agent_name: string; skill_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const result = await apiClient.request<{
          success: boolean;
          message: string;
          skill_name: string;
        }>(
          "POST",
          `/api/agents/${encodeURIComponent(agent_name)}/skills/${encodeURIComponent(skill_name)}`
        );

        return JSON.stringify(result, null, 2);
      },
    },

    // ========================================================================
    // set_agent_skills - Bulk update agent skills
    // ========================================================================
    setAgentSkills: {
      name: "set_agent_skills",
      description:
        "Set all skills for an agent (replaces existing assignments). " +
        "Use this to configure multiple skills at once.",
      parameters: z.object({
        agent_name: z.string().describe("Name of the agent"),
        skills: z.array(z.string()).describe("List of skill names to assign"),
      }),
      execute: async (
        { agent_name, skills }: { agent_name: string; skills: string[] },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const result = await apiClient.request<{
          success: boolean;
          agent_name: string;
          skills_assigned: number;
          skills: string[];
        }>(
          "PUT",
          `/api/agents/${encodeURIComponent(agent_name)}/skills`,
          { skills }
        );

        return JSON.stringify(result, null, 2);
      },
    },

    // ========================================================================
    // sync_agent_skills - Inject skills to running agent
    // ========================================================================
    syncAgentSkills: {
      name: "sync_agent_skills",
      description:
        "Inject all assigned skills into a running agent. " +
        "Copies SKILL.md files to the agent's .claude/skills/ directory. " +
        "Agent must be running. Use this after assigning skills to apply them immediately.",
      parameters: z.object({
        agent_name: z.string().describe("Name of the agent to sync skills to"),
      }),
      execute: async (
        { agent_name }: { agent_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const result = await apiClient.request<SkillInjectionResult>(
          "POST",
          `/api/agents/${encodeURIComponent(agent_name)}/skills/inject`
        );

        if (result.success) {
          return JSON.stringify({
            success: true,
            message: `Successfully injected ${result.skills_injected} skills to agent ${agent_name}`,
            skills_injected: result.skills_injected
          }, null, 2);
        } else {
          return JSON.stringify({
            success: false,
            message: `Injected ${result.skills_injected} skills, ${result.skills_failed} failed`,
            skills_injected: result.skills_injected,
            skills_failed: result.skills_failed,
            results: result.results
          }, null, 2);
        }
      },
    },

    // ========================================================================
    // get_agent_skills - Get skills assigned to an agent
    // ========================================================================
    getAgentSkills: {
      name: "get_agent_skills",
      description:
        "Get the list of skills assigned to an agent.",
      parameters: z.object({
        agent_name: z.string().describe("Name of the agent"),
      }),
      execute: async (
        { agent_name }: { agent_name: string },
        context?: { session?: McpAuthContext }
      ) => {
        const authContext = context?.session;
        const apiClient = getClient(authContext);

        const skills = await apiClient.request<Array<{
          id: number;
          agent_name: string;
          skill_name: string;
          assigned_by: string;
          assigned_at: string;
        }>>(
          "GET",
          `/api/agents/${encodeURIComponent(agent_name)}/skills`
        );

        return JSON.stringify({
          agent_name,
          skill_count: skills.length,
          skills: skills.map(s => ({
            name: s.skill_name,
            assigned_by: s.assigned_by,
            assigned_at: s.assigned_at
          }))
        }, null, 2);
      },
    },
  };
}
